### ISSUE TITLE
DocumentExecutor throwing exception

### ISSUE THREAD

**Issue by TechGeeky** [graphql-dotnet/graphql-dotnet#2635]
@Shane32 We are seeing issue where document executor is throwing exception. Below are the stacktraces we saw in production.

We are still using the `UnhandledExceptionDelegate` delegate as it is.

```
at GraphQLMiddleware.cs:line 214 (this is our "context.Response.StatusCode = statusCode;" line)
   at GraphQL.Execution.ExecutionStrategy.ProcessNodeUnhandledException(ExecutionContext context, ExecutionNode node, Exception ex) in /_/src/GraphQL/Execution/ExecutionStrategy.cs:line 316
   at GraphQL.Execution.ExecutionStrategy.CompleteDataLoaderNodeAsync(ExecutionContext context, ExecutionNode node) in /_/src/GraphQL/Execution/ExecutionStrategy.cs:line 239
   at GraphQL.Execution.ParallelExecutionStrategy.ExecuteNodeTreeAsync(ExecutionContext context, ExecutionNode rootNode) in /_/src/GraphQL/Execution/ParallelExecutionStrategy.cs:line 63
   at GraphQL.Execution.ExecutionStrategy.ExecuteAsync(ExecutionContext context) in /_/src/GraphQL/Execution/ExecutionStrategy.cs:line 31
   at GraphQL.DocumentExecuter.ExecuteAsync(ExecutionOptions options) in /_/src/GraphQL/Execution/DocumentExecuter.cs:line 223
   ```

And this as well - 

```
at GraphQLMiddleware.cs:line 214 (this is our "context.Response.StatusCode = statusCode;" line)
   at GraphQL.DocumentExecuter.ExecuteAsync(ExecutionOptions options) in /_/src/GraphQL/Execution/DocumentExecuter.cs:line 261
at GraphQLMiddleware.cs:line 112
```

So mainly it is coming from either `DocumentExecuter.cs:line 261` or `DocumentExecuter.cs:line 223`. Any idea what could be the reason?

Below is our modified new `ExecuteAsync` method where we added one extra if check in `UnhandledExceptionDelegate` and `HandleErrors` method is just a void method with bunch of sync methods.

```
private async Task ExecuteAsync(HttpContext context, ISchema schema)
{
    var loggingService = context.RequestServices.GetService<ILoggingService>();
    var loggingContext = new LoggingContext(context.Request.Headers);
    context.Items["logRequest"] = new RequestInfo();
    using (var logTimer = new LogTimer(loggingService, loggingContext, "Debug Logs", GetEndpoint(context), "start"))
    {
        try
        {
            var request = Deserialize<GraphQLRequest>(context.Request.Body);
            var result = await _executer.ExecuteAsync(opts =>
            {
                opts.Schema = schema;
                opts.Query = request?.Query;
                opts.OperationName = request?.OperationName;
                opts.Inputs = request?.Variables.ToInputs();
                opts.RequestServices = context.RequestServices;
                opts.UserContext = _settings.BuildUserContext?.Invoke(context);
                opts.ValidationRules = DocumentValidator.CoreRules.Concat(new[] { new InputValidationRule() });
                opts.EnableMetrics = _settings.EnableMetrics;
                if (_settings.EnableMetrics)
                {
                    opts.FieldMiddleware.Use<InstrumentFieldsMiddleware>();
                }
                opts.UnhandledExceptionDelegate = ctx =>
                {
                    if (ctx.Exception is OverflowException of)
                    {
                        throw new ApiDeserializationException(StatusCodes.Status204NoContent, of.Message, null);
                    }
                   HandleErrors(context, ctx.Exception, loggingService, loggingContext);
                };
            });
            await WriteResponseAsync(context, result);
            logTimer.StatusCode = context.Response.StatusCode;
            logTimer.Message = ((RequestInfo)context.Items["logRequest"]).Value;
        }
        catch (Exception ex)
        {
            HandleErrors(context, ex, loggingService, loggingContext);
            logTimer.StatusCode = context.Response.StatusCode;
            logTimer.Message = ((RequestInfo)context.Items["logRequest"]).Value;
        }
    }
}
```

We are using 

```
<PackageReference Include="GraphQL.Server.Ui.Playground" Version="4.2.0" />
<PackageReference Include="GraphQL" Version="3.1.3" />
```

One thing to note is `ExecuteAsync` method is `async` method but "HandleErrors" is not async.

**Comment by TechGeeky** [comment]
@Shane32 Any thoughts on this? I even tried using third option as mentioned [here](https://github.com/graphql-dotnet/graphql-dotnet/issues/2615#issuecomment-892022744) which is local scope and still I see exception being thrown. Any other thoughts which I can try?

**Comment by Shane32** [comment]
I assume you are still getting a NRT exception.  This can only occur if the document executer returns before this unhandled delegate code is run.  My best guess is that perhaps with parallel execution, execution is being terminated before one of the field resolvers gets awaited within the execution strategy.

For example, consider this scenario:

1. A query request arrives with a long running async task and a long running sync task:

```gql
{
  products { id name }  # async task
  categories { id name }  # sync task
}
```

2. 'products' starts executing at ParallelExecutionStrategy.cs:50, which is async
3. 'categories' starts executing at ParallelExecutionStrategy.cs:50, but since it is synchronous, has a delay here
4. Client terminates request during this delay, causing the CancellationToken to be signalled
5. ExecutionStrategy.cs:418 for 'categories' rethrows an OperationCanceledException
6. ExecutionStrategy.cs:54 rethrows the OperationCanceledException and so never awaits the return of the 'ExecuteNodeAsync' method for 'products'.
7. The document executer finishes up and ASP.NET clears the HttpContext
8. ASP.NET also closes the scope, which calls `Close` on the underlying database used by 'products'.
9. The 'products' field resolver throws an ObjectDisposedException due to the database being closed.
10. The `ExecuteNodeAsync` method for 'products' processes the ObjectDisposedException.
10. The exception triggers ProcessNodeUnhandledException at ExecutionStrategy.cs:426
11. The unhandled exception delegate now has a reference to an http context that has already been cleared by ASP.Net

It is possible that this is the reason you are seeing this error, or something similar.  I would consider this to be a bug in the design of GraphQL.NET, as all tasks should be awaited before the ExecutionStrategy finishes.

Could you test a patch?  I think probably this needs to be added above the `finally` on ParallelExecutionStrategy.cs:111:

```cs
catch
{
    try
    {
        await OnBeforeExecutionStepAwaitedAsync(context).ConfigureAwait(false);
    }
    catch
    {
    }
    try
    {
        await Task.WhenAll(currentTasks).ConfigureAwait(false);
    }
    catch
    {
    }
    throw;
}
```

You'll need to duplicate the ParallelExecutionStrategy class and add the patch, then test it in your environment.  If all is good we can include this in 4.6.1.

**Comment by TechGeeky** [comment]
@Shane32 Thanks for the fix. I saw it was merged for 4.6.1 release. When do you think we can get this release 4.6.1 out to public?

**Comment by Shane32** [comment]
Released.
